les/fetcher.go

```diff
@@ -232,11 +244,11 @@ func (f *lightFetcher) unregisterPeer(p *peer) {
 	p.hasBlock = nil
 	p.lock.Unlock()
 
-	f.lock.Lock()
-	defer f.lock.Unlock()
-
 	// check for potential timed out block delay statistics
 	f.checkUpdateStats(p, nil)
+
+	f.lock.Lock()
```

Comment:
Why did you move the locking here? This seems to be the reason why checkUpdateStats has data races, which AFAIK led you to introduce fetcherPeerInfo.lock too. This function should always be called under lightFetcher.lock, it also accesses lightFetcher.peers so fetcherPeerInfo.lock is both insufficient and unnecessary. Once lightFetcher.lock is locked, it protects all fetcherPeerInfo structures too.

Reply:
Yes, it was datarace 
```WARNING: DATA RACE
Write at 0x00c425b2c810 by goroutine 74:
  runtime.mapdelete_fast64()
      /usr/local/go/src/runtime/hashmap_fast.go:834 +0x0
  github.com/ethereum/go-ethereum/les.(*lightFetcher).unregisterPeer()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:252 +0x127
  github.com/ethereum/go-ethereum/les.(*peerSet).Unregister()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/peer.go:601 +0x246
  github.com/ethereum/go-ethereum/les.testOdr()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/odr_test.go:206 +0x8e2
  github.com/ethereum/go-ethereum/les.TestOdrGetBlockLes1()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/odr_test.go:42 +0x56
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c425b2c810 by goroutine 89:
  runtime.mapiternext()
      /usr/local/go/src/runtime/hashmap.go:740 +0x0
  github.com/ethereum/go-ethereum/les.(*lightFetcher).updateMaxConfirmedTd()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:930 +0x23d
  github.com/ethereum/go-ethereum/les.(*lightFetcher).newHeaders()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:647 +0x2fe
  github.com/ethereum/go-ethereum/les.(*lightFetcher).syncLoop()
      /Users/boris/go/src/github.com/ethereum/go-ethereum/les/fetcher.go:222 +0x50b
```

Reply:
I know that you had a data race, that's not what I'm asking. Here you swapped the order of `f.checkUpdateStats(p, nil)` and `f.lock.Lock()` which is clearly wrong because checkUpdateStats was meant to be called while the main fetcher lock is held. Why did you do that? To me it looks like this is the CAUSE of the data race, not a solution to it.